src/Bicep.Core/Modules/TemplateSpecModuleReference.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Text.RegularExpressions;
+using Azure.Deployments.Core.Uri;
+using Bicep.Core.Diagnostics;
+
+namespace Bicep.Core.Modules
+{
+    public class TemplateSpecModuleReference : ModuleReference
+    {
+        private static readonly UriTemplate FullyQualifiedTemplateSpecUriTemplate = new("/{endpoint}/{subscriptionId}/{resourceGroupName}/{templateSpecName}:{version}");
+
+        private static readonly UriTemplate DefaultTemplateSpecUriTemplate = new("/{subscriptionId}/{resourceGroupName}/{templateSpecName}:{version}");
```

Comment:
The format of template spec references is not finalized. I'm planning on discussing it in our next DSL discussion meeting. Some open questions I have:
- Will people prefer the reference string to be a resource ID?
- Should `endpoint` be an optional part of the reference (the current implemenation)? Or does it makes sense to move it to `bicepconfig.json` so people can deploy the same infrastructure to different cloud environments by simply switching bicep config files?